cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -259,223 +252,86 @@ type PaginationOptions struct {
-
-func (id *EventID) Parse(input string) error {
-	parts := strings.SplitN(input, "-", 2)
-	if len(parts) != 2 {
-		return fmt.Errorf("invalid event id %s", input)
-	}
-
-	// Parse the first part (toid)
-	idInt, err := strconv.ParseInt(parts[0], 10, 64) //lint:ignore gomnd
-	if err != nil {
-		return errors.Wrapf(err, "invalid event id %s", input)
-	}
-	parsed := toid.Parse(idInt)
-
-	// Parse the second part (event order)
-	eventOrder, err := strconv.ParseInt(parts[1], 10, 64) //lint:ignore gomnd
-	if err != nil {
-		return errors.Wrapf(err, "invalid event id %s", input)
-	}
-
-	id.ID = &parsed
-	// Subtract one to go from the id to the
-	id.EventOrder = int32(eventOrder) - 1
-
-	return nil
+type eventScanner interface {
+	Scan(eventRange events.Range, f func(events.Cursor, int64, xdr.ContractEvent) bool) error
 }
 
-//lint:ignore gocyclo
-func (a EventStore) GetEvents(request GetEventsRequest) ([]EventInfo, error) {
+func getEvents(scanner eventScanner, request GetEventsRequest) ([]EventInfo, error) {
 	if err := request.Valid(); err != nil {
 		return nil, err
 	}
 
-	finish := toid.AfterLedger(request.EndLedger)
-
-	var results []EventInfo
-
-	// TODO: Use a more efficient backend here. For now, we stream all ledgers in
-	// the range from horizon, and filter them. This sucks.
-	cursor := EventID{
-		ID:         toid.New(request.StartLedger, int32(0), 0),
-		EventOrder: 0,
-	}
+	start := events.Cursor{Ledger: uint32(request.StartLedger)}
+	limit := maxLimit
 	if request.Pagination != nil && request.Pagination.Cursor != "" {
-		if err := cursor.Parse(request.Pagination.Cursor); err != nil {
-			return nil, errors.Wrap(err, "invalid pagination cursor")
-		}
-	}
-	err := a.ForEachTransaction(cursor.ID, finish, func(transaction horizon.Transaction) error {
-		// parse the txn paging-token, to get the transactionIndex
-		pagingTokenInt, err := strconv.ParseInt(transaction.PagingToken(), 10, 64) //lint:ignore gomnd
+		var err error
+		start, err = events.ParseCursor(request.Pagination.Cursor)
 		if err != nil {
-			return errors.Wrapf(err, "invalid paging token %s", transaction.PagingToken())
-		}
-		pagingToken := toid.Parse(pagingTokenInt)
-
-		// For the first txn, we might have to skip some events to get the first
-		// after the cursor.
-		operationCursor := cursor.OperationOrder
-		eventCursor := cursor.EventOrder
-		cursor.OperationOrder = 0
-		cursor.EventOrder = 0
-		if pagingToken.ToInt64() > cursor.ToInt64() {
-			// This transaction is after the cursor, so we need to reset the cursor
-			operationCursor = 0
-			eventCursor = 0
-		}
-
-		var meta xdr.TransactionMeta
-		if err := xdr.SafeUnmarshalBase64(transaction.ResultMetaXdr, &meta); err != nil {
-			// Invalid meta back. Eek!
-			return err
+			return nil, errors.Wrap(err, "invalid pagination cursor")
 		}
-
-		v3, ok := meta.GetV3()
-		if !ok {
-			return nil
+		// increment event index because, when paginating,
+		// we start with the item right after the cursor
+		start.Event++
```

Comment:
Again, should make sure we document this, as I think it might currently say it is inclusive.